data/transactions/verify/txn.go

```diff
@@ -448,3 +456,394 @@ func (w *worksetBuilder) next() (txnGroups [][]transactions.SignedTxn) {
+	// instead of having many smaller batching waiting in the execution pool
+	numberOfExecPoolSeats := verificationPool.GetParallelism() * 2
+
+	sv = &StreamVerifier{
+		seatReturnChan:   make(chan interface{}, numberOfExecPoolSeats),
+		resultChan:       resultChan,
+		stxnChan:         stxnChan,
+		verificationPool: verificationPool,
+		ctx:              ctx,
+		cache:            cache,
+		nbw:              nbw,
+		ledger:           ledger,
+	}
+	for x := 0; x < numberOfExecPoolSeats; x++ {
+		sv.seatReturnChan <- struct{}{}
+	}
+	return sv
+}
+
+// Start is called when the verifier is created and whenever it needs to restart after
+// the ctx is canceled
+func (sv *StreamVerifier) Start() {
+	go sv.batchingLoop()
+}
+
+func (sv *StreamVerifier) batchingLoop() {
+	timer := time.NewTicker(waitForFirstTxnDuration)
+	var added bool
+	var numberOfSigsInCurrent uint64
+	uel := makeUnverifiedElementList(sv.nbw, sv.ledger)
+	for {
+		select {
+		case stx := <-sv.stxnChan:
+			isFirstInBatch := numberOfSigsInCurrent == 0
+			numberOfBatchableSigsInGroup, err := getNumberOfBatchableSigsInGroup(stx.TxnGroup)
+			if err != nil {
+				// wrong number of signatures
+				sv.sendResult(stx.TxnGroup, stx.Context, err)
+				continue
+			}
+
+			// if no batchable signatures here, send this as a task of its own
+			if numberOfBatchableSigsInGroup == 0 {
+				// no point in blocking and waiting for a seat here, let it wait in the exec pool
+				err := sv.addVerificationTaskToThePool(makeSingleUnverifiedElement(sv.nbw, sv.ledger, stx), false)
+				if err != nil {
+					continue // pool is terminated. nothing to report to the txn sender
+				}
+				continue // stx is handled, continue
+			}
+
+			// add this txngrp to the list of batchable txn groups
+			numberOfSigsInCurrent = numberOfSigsInCurrent + numberOfBatchableSigsInGroup
+			uel.elementList = append(uel.elementList, stx)
+			if numberOfSigsInCurrent > txnPerWorksetThreshold {
+				// enough transaction in the batch to efficiently verify
+
+				if numberOfSigsInCurrent > 4*txnPerWorksetThreshold {
+					// do not consider adding more txns to this batch.
+					// bypass the seat count and block if the exec pool is busy
+					// this is to prevent creation of very large batches
+					err := sv.addVerificationTaskToThePool(uel, false)
+					if err != nil {
+						continue // pool is terminated.
+					}
+					added = true
+				} else {
+					added = sv.processBatch(uel)
+				}
+				if added {
+					numberOfSigsInCurrent = 0
+					uel = makeUnverifiedElementList(sv.nbw, sv.ledger)
+					// starting a new batch. Can wait long, since nothing is blocked
+					timer.Reset(waitForFirstTxnDuration)
+				} else {
+					// was not added because no-available-seats. wait for some more txns
+					timer.Reset(waitForNextTxnDuration)
+				}
+			} else {
+				if isFirstInBatch {
+					// an element is added and is waiting. shorten the waiting time
+					timer.Reset(waitForNextTxnDuration)
+				}
+			}
+		case <-timer.C:
```

Comment:
The priority is given to latency, so that txns can be pushed over the network as fast as possible. 
When the txn volume is low, then yes, we will have smaller batches. But this is fine, since the cpu can take more load instead of sitting idle, and not introduce any unnecessary latency.
When the txn load is high, the latency will not suffer because of the batching, and the cpu will have less work to do.

Reply:
Good point. 10 or 20 ms?

Reply:
done. increased to 20ms.